README.md
## TTDX Editor
This repository contains the source code to TTDX Editor, a tool for editing Transport Tycoon Delxuxe (and TTDPatch) saved games and scenarios. It was originally written by Jens Vang Peterson, and has been sporadically updated and maintained by Owen Rudge since 2002. You can read more about the software at https://www.transporttycoon.net/ttdxedit.

## Licence
The source code to TTDX Editor was released in October 2017 under the GNU General Public Licence version 2. Please respect this licence should you choose to fork or re-release this code.

## Source code and development environment
The code is written in Visual Basic 6, and the code currently makes use of a number of third-party ActiveX controls. These are primarily TimoSoft controls (https://www.timosoft-software.de/), which are free of charge.

The code has not in any way been tidied up for public release. The last released version of TTDX Editor was version 1.20; the current master represents a development version of the code, which included work adding support for Unicode controls (for translation/text entry purposes, where appropriate). It may be buggy or incomplete - it may not even compile. The documentation that gets shipped with the Setup program has NOT been updated to indicate that the code is now licensed under the GPL - if you happen to want to release a new version (or have code that you'd like me to release as part of a new version), please do ensure the documentation is updated appropriately.

If you have any queries, feel free to drop me (Owen Rudge) an e-mail and I will be happy to try to help.

## C# rewrite
A few years ago I began attempting to rewrite TTDX Editor in C#. This is a work-in-progress, but should sufficient progress be made, I will update this repository with the new code.

changes.txt
TTDX Editor Changes
===================

Version 1.20 (12/03/2012)
-------------------------

- Marked saved games with basic record of what was changed (24/05/2004)
- Update MODIFIED.EXE to list game modifications (19/06/2004)
- Assorted cosmetic changes (10/03/2012)
- Run nicely on modern versions of Windows, and with limited privileges (10/03/2012)
- Update SGM plug-in with support for 2.3 registration system (12/03/2012)
- Setup program improvements; don't bundle WSH or DCOM95 (12/03/2012)
- Add "Check for Updates" feature (12/03/2012)

Version 1.10.017 (23/05/2004)
-----------------------------

- Fixed overflow bug with finance window (15/05/2004)
- Added multiple currency support (15/05/2004)
- Moved registry entries from JVP to Owen Rudge, made editor remember last window position (22/05/2004)
- By Oskar Eisemuth: Fixed saving bug with TTDPatch 2.0.1 alpha 27 saved games (23/05/2004)
- Updated SGM plug-in for new registry entries (23/05/2004)

Version 1.10.016 (26/12/2003)
-----------------------------

- Added Finances window (14/10/2003)
- Changed fonts to Tahoma (14/10/2003)
- Fixed vehicle list problem (14/10/2003)
- Added workaround for bug caused by abandonedroads (chunk size specified was different to
  actual chunk size). This is fixed in TTDPatch 2.0.1 alpha 10. (26/12/2003)

Version 1.10.015 (11/10/2003)
-----------------------------

- Rewrote message and fixed XP themes problem in File Types window (05/10/2003)
- Made a few cosmetic changes here and there (05/10/2003)
- Fixed bug caused by relocation of &H24CBC to &H44CB8 from TTDPatch 2.0 onwards (11/10/2003)

Version 1.10.014 (30/08/2003)
-----------------------------

- Add buildings around a station
- Techinfo in "free" window, can be resized
- Fixed a very bad bug, would often cause games to crash.
- Date sorting in select game fixed
- Cleanout settings option
- SGM Plugin setup redone, should work with v2.1 also

- Owen Rudge became maintainer (10/07/2003)
- Added support for games created with TTDPatch 1.9.1 alpha 32 or newer (10/07/2003)
- Updated some windows, replaced About dialog (10/07/2003)
- 'Marked' files that have been edited (12/07/2003)
- Fixed overflow error with high morevehicles values in Vehicle Editor (14/07/2003)
- Added Save As dialog (16/07/2003)
- Got everything ready for release (30/08/2003)
- Added code to support TTDPatch 2.0 beta 5.1 MG or higher (13/09/2003)

Version 1.10.013 (05/05/2002)
-----------------------------
 
- Vehicle Editor
- Own AI roads
- Own city bridges
- Techinfo rewritten (again)
- Menu navigation keys (Thanks to Owen Rudge for reminding me)
- Fixed some problems with file associations

Version 1.00.012 (26/04/2002)
-----------------------------
 
- Stations can be filtered by owner and sorted
- Added some shortcut keys
- More speed when loading and saving
- New map drawing procedure, should be faster
- More sorting and filtering options in file selector
- Plugin for Saved Game Manager by Owen Rudge
- Fixed citynames in the industry section
- Minor fixes in the about
 
Version 1.00.010 (20/04/2002)
-----------------------------

- Quick fix: A very stupid bug in the path detection.
 
Version 1.00.009 (20/04/2002)
-----------------------------

- Decode game-generated city names
- Remove HQ for any player
- New About Window
- New status bar
- Fixed X-cord on large/extreme map
- Fixed crash in techinfo if minimized
 
Version 1.00.008 (16/04/2002)
-----------------------------

- Change cargo ratings and amount at stations
- The map window working, 3 sizes or turned off
- Filetypes association
- Minimize Industries production macro
- Remove Trees
- City ratings now set with a slider
- Re-selecting an open editor now brings it to front
 
Version 1.00.004 (06/04/2002)
-----------------------------

- Change player money and debt
- Change city ratings and def. name
- Load and Save uncompressed now working
- Command line option to directly load a game
- Techinfo moved to seperate window, can be hidden
- Fixed: Citynames wouldn't show in games saved without the TTDPatch
- Fixed: Multiplayer savegames (sv2) were handled as they were scenarios
- Fixed: Some changed values would not be included in a save
 
Version 1.00.003 (30/03/2002, first public alpha)
-------------------------------------------------
 
- All known industrial values now included
- Industrial types and cargo for Toyland climate included
- Well-known values are shown in different color in Tech-info
- File requester shows the tycoon name of the games
- File requester settings are stored
- Header checksum re-calculated when saving
- Supports drag-drop of savegames
 
Version 1.00.001 (not public)
-----------------------------
 
- Everything


Modified/modified.c
/************************************************************/
/* TTDX Editor - MODIFIED.EXE Application                   */
/* Version 1.20                                             */
/*                                                          */
/* Copyright (c) Owen Rudge 2003-2012. All Rights Reserved. */
/************************************************************/
/* Revision History:
 *
 * 13/09/2003: Updated to work with TTDPatch 2.0 beta
 *             5.1 MG or higher (orudge)
 *
 * 19/06/2004: Updated to add support for listing modifications (orudge)
 */
 
#include <stdio.h>
#include <string.h>

char version_greater(char ma1, char mi1, char r1, char ma2, char mi2, char r2);
void list_modifications(char mod);

void print_usage(char *appname)
{
	printf("TTDX Editor Game Modification Checker\n");
	printf("Version 1.20\n");
	printf("\n");
	printf("Usage: %s filename\n", appname);
	printf("\n");
	printf("This utility will tell you whether a game has been modified with TTDX\n");
	printf("Editor 1.10.014 or later.\n");
	printf("\n");
	printf("'filename' should be an uncompressed ('big') Transport Tycoon Deluxe\n");
	printf("saved game or scenario (eg, created by 'Save Uncompressed' in TTDX\n");
	printf("Editor, or decoded with SV1Codec)\n");
}
	
int main(int argc, char *argv[])
{
	FILE *in;
	char major, minor, revision, modifications;
	long newformat;
	
	if (argc != 2)
	{
		if (argc == 1)
		{
			print_usage(argv[0]);
			return(1);
		}
		else if (argc == 3)
		{
			if (stricmp(argv[1], "/r") == 0)
			{
				in = fopen(argv[2], "wb");

				if (in == NULL)
				{
					printf("Error: Unable to open %s for writing!\n", argv[1]);
					return(1);
				}

				if (fseek(in, 0x44CB4, SEEK_SET) != 0)
				{
					printf("Error: Unable to seek to 0x44CB4. File may not be a decompressed TTD saved game.\n");
					fclose(in);
					return(1);
				}

				fread(&newformat, 4, 1, in);

				if (newformat != 0x70445454)
				{
					if (fseek(in, 0x24CCB, SEEK_SET) != 0)
					{
						printf("Error: Unable to seek to 0x24CCB. File may not be a decompressed TTD saved game.\n");
						fclose(in);
						return(1);
					}
	
					fputc(0, in);
					fputc(0, in);
					fputc(0, in);

					printf("This saved game has had its modification signature reset.\n");
					return(0);
				}
				else
				{
					if (fseek(in, 0x44BBA, SEEK_SET) != 0)
					{
						printf("Error: Unable to seek to 0x44BBA. File may not be a decompressed TTD saved game.\n");
						fclose(in);
						return(1);
					}
	
					fputc(0, in);
					fputc(0, in);
					fputc(0, in);

					printf("This saved game has had its modification signature reset.\n");
					return(0);
				}
			}									
			else
			{
				print_usage(argv[0]);
				return(1);
			}
		}
	}
	else
	{
		in = fopen(argv[1], "rb");

		if (in == NULL)
		{
			printf("Error: Unable to open %s for reading!\n", argv[1]);
			return(1);
		}

		if (fseek(in, 0x44CB4, SEEK_SET) != 0)
		{
			printf("Error: Unable to seek to 0x44CB4. File may not be a decompressed TTD saved game.\n");
			fclose(in);
			return(1);
		}

		fread(&newformat, 4, 1, in);
		
		if (newformat != 0x70445454)
		{
			if (fseek(in, 0x24CCB, SEEK_SET) != 0)
			{
				printf("Error: Unable to seek to 0x24CCB. File may not be a decompressed TTD saved game.\n");
				fclose(in);
				return(1);
			}

			major = fgetc(in);
			minor = fgetc(in);
			revision = fgetc(in);
			modifications = fgetc(in);
			
			if (major == 0 && minor == 0 && revision == 0)
				printf("This saved game has not been modified by TTDX Editor 1.10.0014 or later.\n");
			else
			{
				printf("This saved game has been modified by TTDX Editor %d.%02d.%04d.\n", major, minor, revision);

				if (version_greater(major, minor, revision, 1, 10, 18))
					list_modifications(modifications);
			}
		}
		else
		{
			if (fseek(in, 0x44BBA, SEEK_SET) != 0)
			{
				printf("Error: Unable to seek to 0x44BBA. File may not be a decompressed TTD saved game.\n");
				fclose(in);
				return(1);
			}

			major = fgetc(in);
			minor = fgetc(in);
			revision = fgetc(in);
			modifications = fgetc(in);

			if (major == 0 && minor == 0 && revision == 0)
				printf("This saved game has not been modified by TTDX Editor 1.10.0014 or later.\n");
			else
			{
				printf("This saved game has been modified by TTDX Editor %d.%02d.%04d.\n", major, minor, revision);

				if (version_greater(major, minor, revision, 1, 10, 18))
					list_modifications(modifications);
			}
		}

		fclose(in);
	}

	return(0);
}

char version_greater(char ma1, char mi1, char r1, char ma2, char mi2, char r2)
{
	if (ma1 > ma2)
		return(1);

	if (ma1 == ma2)
	{
		if (mi1 > mi2)
			return(1);

		if (mi1 == mi2)
		{
			if (r1 > r2)
				return(1);

			if (r1 == r2)
				return(1);
		}
	}

	return(0);
}

void list_modifications(char mod)
{
	puts("\nThe following modifications have been made to the saved game:");

	if (mod & 1)
		puts("- Player data updated");

	if (mod & 2)
		puts("- City data updated");
	
	if (mod & 4)
		puts("- Industry data updated");

	if (mod & 8)
		puts("- Stations data updated");

	if (mod & 16)
		puts("- Vehicles data updated");

	if (mod & 32)
		puts("- Terrain data updated");

	if (mod & 64)
		puts("- Other unknown data updated");

	if (mod & 128)
		puts("- Other unknown data updated");
}
